### PR TITLE
DiscoverConfigs: Ignore dirs and backup files and rewrite tests.

### PR DESCRIPTION
**Pull request by magnusbaeck**

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -53,8 +54,17 @@ func DiscoverConfigs(file_or_directory string) (files []string, err error) {
...
 		}
-		for _, filename := range entries {
-			files = append(files, path.Join(file_or_directory, filename.Name()))
+		for _, file := range entries {
+			fullPath := path.Join(file_or_directory, file.Name())
+			if file.IsDir() {
+				emit("Directory found while looking for configuration files, skipping: %s", fullPath)
+				continue
+			}
+			if strings.HasSuffix(file.Name(), "~") {
```
Comment (jordansissel) [inline config.go]: What uses ~ for a backup file? Vim uses `.my-file-name.swp` for most defaults I've seen.

**Comment by jordansissel** [comment]
I forgot I did some research on this before: 
https://github.com/elastic/logstash/issues/2271#issuecomment-67690459

I've copied the full comment for easier reading:

---

I like the idea, but I"m not sold on the proposed solution. What editor uses '~' suffix for backup files? Emacs? Vim uses `.<Filename>.swp` by default, although some users configure backup files to go to an entirely different directory and wouldn't be impacted by this problem.

Tested on my OSX 10.10 laptop:
* vim with *no* configuration loaded: `.some_file_name.swp`
* emacs (default config): when I ran it, added some weird `.#<filename>` symlink pointing at 'user@hostname.pid'
* emacs: after quitting, there was a file named `#<filename>#` in the current directory
* atom (default config): no backup file or other state file was found in the current directory. I have an ancient version of Atom that I never use so I'm not sure if the behavior has changed in the past year.

I haven't tested any other editors, but this behavior seems hardly uniform :(

I'd be open to maybe having the default be more explicit in what filenames are accepted. Perhaps only files matching /[0-9A-Za-z_-]+/ would be ok? This would ignore any dot files and any weird local state files created by the editors I've tested so far.

If you use a different  editor or an editor tested above with a different behavior, please update this ticket with what backup or local state files are created by your editor so we can try to generalize a pattern here.

---

**Comment by jordansissel** [comment]
For clarity, I don't think we need an exhaustive solution, just trying to see what short gaps we can cross earlier. If we only support ignoring files ending in `~` to start, that's ok. :)

**Comment by avenjamin** [comment]
I've seen the ~ suffix used on nano backup files. 
 
Ben

**Comment by magnusbaeck** [comment]
> For clarity, I don't think we need an exhaustive solution, just trying to see what short gaps we can cross earlier. If we only support ignoring files ending in ~ to start, that's ok. :)

As per my comment in the discussion in the other bug, ignoring `.*`, `#*`, and `*~` should take care of the vast majority of weird and unwanted files with a very low risk of false positives. If you're fine with that I can update this PR and prepare PR(s) for Logstash's config reader and probably also the grok pattern reader.

Or, we follow the suggestion in https://github.com/elastic/logstash/issues/1831 and only read *.conf, but that change wouldn't be backwards compatible.

**Comment by magnusbaeck** [comment]
Oops, PR #311 has already been filed to skip directories. However, I think this one cleans up the tests in a nice way.

**Comment by magnusbaeck** [comment]
Ping. If this PR isn't to your liking, please make a decision about what to do with the backup files and what not that continue screw things up for people.

**Comment by magnusbaeck** [comment]
LSF is dead and Filebeat doesn't appear to support conf.d-style directoriies, so there's nothing much to migrate to Filebeat. Closing the PR.

**Comment by ruflin** [comment]
@magnusbaeck Filebeat still supports a config_dir directory with additional config files. It limits the files to ending with `.yml` but it doesn't check how the files start: https://github.com/elastic/beats/blob/master/filebeat/config/config.go#L123

**Pull request closed without merging**